src/gp_doctest.cpp

```diff
@@ -2760,6 +2760,188 @@ TEST_CASE("TPEngine: Branch Length Optimization") {
+        std::ofstream file_out;
+        const std::string temp_newick_path_1 = "_ignore/temp_1.newick";
+        file_out.open(temp_newick_path_1);
+        file_out << inst_1.GetTPEngine().ToNewickOfTopTrees() << std::endl;
+        file_out.close();
+        const std::string temp_newick_path_2 = "_ignore/temp_1.newick";
+        file_out.open(temp_newick_path_2);
+        const auto tree_map = BuildTopTreeMapViaBruteForce(inst_1);
+        for (const auto& tree : tree_map) {
+          file_out << inst_1.GetDAG().TreeToNewickTopology(tree) << std::endl;
+        }
+        file_out.close();
+
+        auto inst_2 = GPInstanceOfFiles(inst_1.GetFastaSourcePath(), temp_newick_path_2,
+                                        "_ignore/mmapped_pv.data1");
+        inst_2.MakeTPEngine();
+        inst_2.MakeNNIEngine();
+        return TPEngine::Compare(inst_1.GetTPEngine(), inst_2.GetTPEngine(), false);
+      };
+
+  auto inst_1 =
+      GPInstanceOfFiles(fasta_path, newick_path_1, "_ignore/mmapped_pv.data1");
+  auto inst_2 =
+      GPInstanceOfFiles(fasta_path, newick_path_2, "_ignore/mmapped_pv.data2");
+  inst_1.MakeTPEngine();
+  inst_1.MakeNNIEngine();
+  auto& tp_engine = inst_1.GetTPEngine();
+  auto& nni_engine = inst_1.GetNNIEngine();
+  inst_2.MakeTPEngine();
+  inst_2.MakeNNIEngine();
+
+  CHECK_MESSAGE(
+      TPEngine::Compare(inst_1.GetTPEngine(), inst_1.GetTPEngine(), false) == 0,
+      "TPEngines not equal to self.");
+  CHECK_MESSAGE(SubsplitDAG::Compare(inst_1.GetDAG(), inst_2.GetDAG(), true) == 0,
+                "DAGs formed from shuffled Newicks not equal.");
+  CHECK_MESSAGE(
+      TPEngine::Compare(inst_1.GetTPEngine(), inst_2.GetTPEngine(), true) != 0,
+      "TPEngines formed from shuffled Newicks are incorrectly equal.");
+  CHECK_MESSAGE(BuildSpanningNewickAndCompareNewDAG(inst_1) == 0,
+                "DAG built from Spanning Newick not equal to the DAG that build it "
+                "(before adding NNIs).");
+  CHECK_MESSAGE(
+      BuildTopTreeNewickAndCompareNewTPEngine(inst_1) == 0,
+      "TPEngine built from Top Tree Newick not equal to the TPEngine that build it "
+      "(before adding NNIs).");
+  CHECK_MESSAGE(BuildTopTreesTest(inst_1),
+                "Top Newick Trees match trees built via brute force match TPEngine "
+                "method (before adding NNIs).");
+
+  size_t optimization_count = 2;
+  tp_engine.GetLikelihoodEvalEngine().SetOptimizationMaxIteration(optimization_count);
+  tp_engine.GetLikelihoodEvalEngine().BranchLengthOptimization(false);
+  nni_engine.SetTPLikelihoodCutoffFilteringScheme(0.0);
+  nni_engine.SetTopNScoreFilteringScheme(1);
+  nni_engine.SetReevaluateRejectedNNIs(true);
+  nni_engine.RunInit(true);
+
+  for (size_t iter = 0; iter < 5; iter++) {
+    nni_engine.RunMainLoop(true);
+    // nni_engine.GetDAG().PrintStorage();
+
+    // if (iter == 7) break;
+    // Issue #xxx: this creates an unknown problem on iteration 7.
```

Comment:
Let's make an issue if this is a problem...

Reply:
Added issue.